include/tvm/runtime/container/array.h

```diff
@@ -574,6 +574,44 @@ class Array : public ObjectRef {
   /*! \return The underlying ArrayNode */
   ArrayNode* GetArrayNode() const { return static_cast<ArrayNode*>(data_.get()); }
 
+  /*!
+   * \brief Helper function to apply a map function onto the array.
+   *
+   * \param fmap The transformation function T -> U.
+   *
+   * \tparam F The type of the mutation function.
+   *
+   * \tparam U The type of the returned array, inferred from the
+   * return type of F.  If overridden by the user, must be something
+   * that is convertible from the return type of F.
+   *
+   * \note This function performs copy on write optimization.  If
+   * `fmap` returns an object of type `T`, and all elements of the
+   * array are mapped to themselves, then the returned array will be
+   * the same as the original, and reference counts of the elements in
+   * the array will not be incremented.
+   *
+   * \return The transformed array.
+   */
+  template <typename F, typename U = std::invoke_result_t<F, T>>
+  Array<U> Map(F fmap) const {
```

Comment:
I was curious if we could unify and migrate `MutateByApply` into `Map`?

Reply:
Possibly, and that would allow for avoiding copies in a few additional cases (e.g. map from `T` to `Optional<T>`, or to a superclass of `T`) that aren't currently handled.  I'll take a quick stab at it and see if I can unify the two.

Reply:
Thank you on the suggestion, and it ended up being much cleaner that way.  Both `Map` and `MutateByApply` are now implemented in terms of the same underlying helper function.  The helper function applies both the mutate-in-place and copy-on-write optimizations, with `if constexpr` type checks to avoid attempting the optimization if they wouldn't be possible.